File: docs/i18n.js
/** English-first i18n — syncs with in-game WASM locale via set_web_locale. */
(function () {
  const STORAGE_KEY = 'qt-lang';
  const DEFAULT = 'en';

  const T = {
    fr: {
      'meta.title': 'Quantum Tetris',
      'lang.group': 'Langue',
      'lang.fr': 'Français',
      'lang.en': 'English',
      'home.title': 'Quantum Tetris',
      'home.subtitle': 'Tetris — événements stochastiques via circuits quantiques',
      'home.lead':
        'Règles Tetris habituelles. Pièce, état d’apparition, cadence, bonus : chaque événement stochastique exécute un circuit prédéfini ; les bits mesurés déterminent l’effet en jeu.',
      'home.stack':
        '<strong>Architecture.</strong> Bevy (Rust) → WASM dans le navigateur. <a href="https://github.com/Renmusxd/RustQIP">RustQIP</a> (<code>RustQipBackend</code>) simule le statevector par défaut en local et en ligne.',
      'home.principle':
        '<strong>Le joueur :</strong> ← → ↑ ↓ et Espace. <strong>Le jeu :</strong> tout le reste.',
      'home.quantumTitle': 'Mécaniques',
      'home.label.game': 'En jeu',
      'home.label.quantum': 'Circuit',
      'home.moment.piece.title': 'Pièce en cours & « suiv. »',
      'home.moment.piece.game': 'Détermine la forme active et l\'aperçu suivant.',
      'home.moment.piece.quantum':
        '<code>quantum-teleportation-gate-v1</code> (×2) — mesure de Bell inspirée de la téléportation ; les bits mesurés fixent la famille (I, O, T…) et la variante.',
      'home.moment.brain.title': 'Rotation & colonne',
      'home.moment.brain.game': 'Fixe l\'orientation à l\'apparition et la colonne d\'entrée.',
      'home.moment.brain.quantum':
        '<code>imp-brain-v1</code> — 2 qubits mesurés → rotation (0–3) et colonne d\'apparition.',
      'home.moment.hunter.title': 'Cadence de chute',
      'home.moment.hunter.game': 'Intervalle entre deux descentes ; diminue avec le niveau.',
      'home.moment.hunter.quantum':
        '<code>enemy-profile-hunter-v1</code> — 2 qubits mesurés → intervalle de chute pour la pièce en cours.',
      'home.moment.observe.title': 'Espace — chute forcée',
      'home.moment.observe.game': 'Pose immédiate ; bonus de score, parfois une ligne supplémentaire.',
      'home.moment.observe.quantum':
        '<code>observation-pulse-v1</code> — mesure à la pose forcée ; les bits choisissent le bonus.',
      'home.moment.line.title': 'Ligne complétée',
      'home.moment.line.game': 'Multiplicateur de points ×1 à ×4 selon le tirage.',
      'home.moment.line.quantum':
        '<code>q-shard-stabilizer-v1</code> — après effacement d\'une ligne ; les bits fixent le multiplicateur (×1–×4).',
      'home.circuitsLink': 'Référence circuits',
      'home.github': 'GitHub',
      'home.readme': 'Build local',
      'play.starting': 'Chargement…',
      'play.downloading': 'Téléchargement du jeu…',
      'play.initializing': 'Démarrage du moteur…',
      'play.hint': 'Premier chargement volumineux. Attendre la fin du téléchargement.',
      'error.wasmJs':
        'Bundle JavaScript WASM introuvable — lancez ./scripts/build_wasm.sh puis servez docs/.',
      'error.wasmExports': 'Le bundle WASM ne contient pas les exports attendus.',
      'error.wasmNetwork':
        'Téléchargement du jeu interrompu. Vérifiez la connexion puis rechargez la page.',
      'error.wasmHttp':
        'Le binaire WASM n’est pas publié à l’URL attendue. Le workflow Pages n’a probablement pas terminé ou l’artefact est incomplet.',
      'error.wasmNotBinary':
        'L’URL du WASM répond, mais ne renvoie pas un binaire WebAssembly. GitHub Pages sert peut-être une page HTML d’erreur ou un cache incomplet.',
      'error.wasmInit':
        'Le binaire WASM est téléchargé, mais le navigateur ne peut pas l’initialiser. Rechargez sans cache ; si ça persiste, le JS et le WASM publiés ne correspondent peut-être pas.',
      'error.wasmUnknown': 'Chargement WASM impossible.',
      'error.detail': 'Détail',
      'error.canvas': 'Le canvas n\'a pas démarré — recharger la page.',
      'error.run': 'Erreur au lancement du jeu — voir la console.',
    },
    en: {
      'meta.title': 'Quantum Tetris',
      'lang.group': 'Language',
      'lang.fr': 'Français',
      'lang.en': 'English',
      'home.title': 'Quantum Tetris',
      'home.subtitle': 'Tetris — stochastic events from quantum circuits',
      'home.lead':
        'Standard Tetris rules. Piece, spawn state, cadence, bonuses: each stochastic event runs a predefined circuit; measured bits drive the in-game effect.',
      'home.stack':
        '<strong>Architecture.</strong> Bevy (Rust) → WASM in the browser. <a href="https://github.com/Renmusxd/RustQIP">RustQIP</a> (<code>RustQipBackend</code>) simulates the statevector by default locally and online.',
      'home.principle':
        '<strong>The player:</strong> ← → ↑ ↓ and Space. <strong>The game:</strong> everything else.',
      'home.quantumTitle': 'Mechanics',
      'home.label.game': 'In-game',
      'home.label.quantum': 'Circuit',
      'home.moment.piece.title': 'Active piece & “next”',
      'home.moment.piece.game': 'Sets the active shape and the next preview.',
      'home.moment.piece.quantum':
        '<code>quantum-teleportation-gate-v1</code> (×2) — teleportation-inspired Bell measurement; measured bits pick the family (I, O, T…) and variant.',
      'home.moment.brain.title': 'Rotation & column',
      'home.moment.brain.game': 'Sets spawn orientation and entry column.',
      'home.moment.brain.quantum':
        '<code>imp-brain-v1</code> — 2 measured qubits → rotation (0–3) and spawn column.',
      'home.moment.hunter.title': 'Drop cadence',
      'home.moment.hunter.game': 'Interval between grid steps; decreases as level rises.',
      'home.moment.hunter.quantum':
        '<code>enemy-profile-hunter-v1</code> — 2 measured qubits → drop interval for the active piece.',
      'home.moment.observe.title': 'Space — hard drop',
      'home.moment.observe.game': 'Instant lock; score bonus, sometimes one extra line.',
      'home.moment.observe.quantum':
        '<code>observation-pulse-v1</code> — measure on hard drop; bits select the score bonus.',
      'home.moment.line.title': 'Line clear',
      'home.moment.line.game': 'Score multiplier ×1 to ×4 from the draw.',
      'home.moment.line.quantum':
        '<code>q-shard-stabilizer-v1</code> — after a line clear; bits set the score multiplier (×1–×4).',
      'home.circuitsLink': 'Circuit reference',
      'home.github': 'GitHub',
      'home.readme': 'Local build',
      'play.starting': 'Loading…',
      'play.downloading': 'Downloading game…',
      'play.initializing': 'Starting engine…',
      'play.hint': 'Large initial download. Wait for the progress bar to finish.',
      'error.wasmJs':
        'WASM JavaScript bundle missing — run ./scripts/build_wasm.sh, then serve docs/.',
      'error.wasmExports': 'WASM bundle is present but does not expose the expected game entry points.',
      'error.wasmNetwork':
        'Game download was interrupted. Check the connection, then reload the page.',
      'error.wasmHttp':
        'The WASM binary is not published at the expected URL. The Pages workflow may still be running or the artifact is incomplete.',
      'error.wasmNotBinary':
        'The WASM URL responded, but it did not return a WebAssembly binary. GitHub Pages may be serving an HTML error page or an incomplete cache.',
      'error.wasmInit':
        'The WASM binary downloaded, but the browser could not initialize it. Hard-refresh the page; if it persists, the published JS and WASM may not match.',
      'error.wasmUnknown': 'WASM failed to load.',
      'error.detail': 'Detail',
      'error.canvas': 'Canvas did not start — reload the page.',
      'error.run': 'Game failed to start — see console.',
    },
  };

  function getLocale() {
    try {
      const stored = localStorage.getItem(STORAGE_KEY);
      return stored === 'fr' ? 'fr' : DEFAULT;
    } catch {
      return DEFAULT;
    }
  }

  function t(key) {
    const dict = T[getLocale()];
    return dict?.[key] ?? key;
  }

  function notifyGame(lang) {
    const l = lang === 'en' ? 'en' : 'fr';
    if (typeof window.__qtSetWebLocale === 'function') {
      window.__qtSetWebLocale(l);
    } else if (typeof window.__qtQueueLocale === 'function') {
      window.__qtQueueLocale(l);
    }
  }

  function setLocale(lang) {
    const l = lang === 'en' ? 'en' : 'fr';
    try {
      localStorage.setItem(STORAGE_KEY, l);
    } catch {
      /* private mode */
    }
    apply();
    notifyGame(l);
    window.dispatchEvent(new CustomEvent('qt-locale', { detail: l }));
  }

  function apply() {
    const lang = getLocale();
    document.documentElement.lang = lang;
    const dict = T[lang];
    document.querySelectorAll('[data-i18n]').forEach((el) => {
      const key = el.getAttribute('data-i18n');
      const val = dict[key];
      if (val == null) return;
      const attr = el.getAttribute('data-i18n-attr');
      if (attr) {
        el.setAttribute(attr, val);
      } else if (el.hasAttribute('data-i18n-html')) {
        el.innerHTML = val;
      } else {
        el.textContent = val;
      }
    });
    document.title = dict['meta.title'] || document.title;
    document.querySelectorAll('.lang-btn[data-lang]').forEach((btn) => {
      const active = btn.getAttribute('data-lang') === lang;
      btn.classList.toggle('active', active);
      btn.setAttribute('aria-pressed', active ? 'true' : 'false');
    });
    const group = document.querySelector('.lang-flags');
    if (group && dict['lang.group']) {
      group.setAttribute('aria-label', dict['lang.group']);
    }
  }

  window.qtI18n = { getLocale, setLocale, apply, t };

  function bindLangButtons() {
    document.querySelectorAll('.lang-btn[data-lang]').forEach((btn) => {
      btn.addEventListener('click', () => {
        setLocale(btn.getAttribute('data-lang'));
      });
    });
  }

  if (document.readyState === 'loading') {
    document.addEventListener('DOMContentLoaded', () => {
      apply();
      bindLangButtons();
    });
  } else {
    apply();
    bindLangButtons();
  }
})();

File: crates/game/src/i18n.rs
//! UI strings and in-game circuit explanations (English default on web).

use bevy::prelude::*;
#[cfg(all(feature = "wasm", target_arch = "wasm32"))]
use std::sync::atomic::{AtomicBool, AtomicU8, Ordering};

#[derive(Resource, Clone, Copy, PartialEq, Eq, Default, Debug)]
pub enum Locale {
    #[default]
    En,
    Fr,
}

#[cfg(all(feature = "wasm", target_arch = "wasm32"))]
static WEB_LOCALE_DIRTY: AtomicBool = AtomicBool::new(false);
#[cfg(all(feature = "wasm", target_arch = "wasm32"))]
static WEB_LOCALE: AtomicU8 = AtomicU8::new(0);

impl Locale {
    pub fn toggle(self) -> Self {
        match self {
            Self::Fr => Self::En,
            Self::En => Self::Fr,
        }
    }

    pub fn toggle_label(self) -> &'static str {
        match self {
            Self::Fr => "(en)",
            Self::En => "(fr)",
        }
    }

    #[cfg(all(feature = "wasm", target_arch = "wasm32"))]
    pub fn from_web_storage() -> Self {
        web_sys::window()
            .and_then(|w| w.local_storage().ok())
            .flatten()
            .and_then(|s| s.get_item("qt-lang").ok())
            .flatten()
            .filter(|v| v == "fr")
            .map(|_| Self::Fr)
            .unwrap_or(Self::En)
    }
}

#[cfg(all(feature = "wasm", target_arch = "wasm32"))]
pub fn push_web_locale(lang: &str) {
    WEB_LOCALE.store(u8::from(lang == "fr"), Ordering::Relaxed);
    WEB_LOCALE_DIRTY.store(true, Ordering::Relaxed);
}

#[cfg(all(feature = "wasm", target_arch = "wasm32"))]
pub fn sync_web_locale(mut locale: ResMut<Locale>) {
    if WEB_LOCALE_DIRTY.swap(false, Ordering::Relaxed) {
        *locale = match WEB_LOCALE.load(Ordering::Relaxed) {
            1 => Locale::Fr,
            _ => Locale::En,
        };
    }
}

pub fn initial_locale() -> Locale {
    #[cfg(all(feature = "wasm", target_arch = "wasm32"))]
    {
        Locale::from_web_storage()
    }
    #[cfg(not(all(feature = "wasm", target_arch = "wasm32")))]
    {
        Locale::default()
    }
}

/// Gameplay moment that triggers one or more quantum circuits.
#[derive(Clone, Copy, PartialEq, Eq, Default, Debug)]
pub enum GameplayMoment {
    #[default]
    None,
    Spawn,
    Observe,
    LineClear,
    GameOver,
}

pub fn circuit_explain(locale: Locale, moment: GameplayMoment) -> &'static str {
    match (locale, moment) {
        (Locale::Fr, GameplayMoment::Spawn) => {
            "Apparition - 3 circuits:\n\
             - quantum-teleportation-gate-v1: tirage 3 bits de la piece active\n\
             - imp-brain-v1: rotation et colonne\n\
             - enemy-profile-hunter-v1: intervalle de chute"
        }
        (Locale::En, GameplayMoment::Spawn) => {
            "Spawn - 3 circuits:\n\
             - quantum-teleportation-gate-v1: 3-bit active-piece draw\n\
             - imp-brain-v1: rotation and column\n\
             - enemy-profile-hunter-v1: drop interval"
        }
        (Locale::Fr, GameplayMoment::Observe) => {
            "Espace - observation-pulse-v1 (2 qubits):\n\
             mesure a la pose forcee -> bonus de score (parfois une ligne bonus)"
        }
        (Locale::En, GameplayMoment::Observe) => {
            "Space - observation-pulse-v1 (2 qubits):\n\
             measure on hard drop -> score bonus (sometimes an extra line)"
        }
        (Locale::Fr, GameplayMoment::LineClear) => {
            "Ligne effacee - q-shard-stabilizer-v1 (2 qubits):\n\
             mesure apres effacement -> multiplicateur de points (x1 a x4)"
        }
        (Locale::En, GameplayMoment::LineClear) => {
            "Line clear - q-shard-stabilizer-v1 (2 qubits):\n\
             post-clear stabilizer -> score multiplier (x1 to x4)"
        }
        (Locale::Fr, GameplayMoment::GameOver) => {
            "Partie terminee - plus de place pour la nouvelle piece."
        }
        (Locale::En, GameplayMoment::GameOver) => "Game over - no room for the next piece.",
        (_, GameplayMoment::None) => "-",
    }
}

pub fn mode_classic(locale: Locale) -> &'static str {
    match locale {
        Locale::Fr => "CLASSIQUE",
        Locale::En => "CLASSIC",
    }
}

pub fn mode_rustqip(locale: Locale) -> &'static str {
    match locale {
        Locale::Fr => "RUSTQIP",
        Locale::En => "RUSTQIP",
    }
}

pub fn hint_move(locale: Locale) -> &'static str {
    match locale {
        Locale::Fr => "BOUGER",
        Locale::En => "MOVE",
    }
}

pub fn hint_rotate(locale: Locale) -> &'static str {
    match locale {
        Locale::Fr => "TOURNER",
        Locale::En => "ROTATE",
    }
}

pub fn hint_faster(locale: Locale) -> &'static str {
    match locale {
        Locale::Fr => "VITE",
        Locale::En => "FASTER",
    }
}

pub fn hint_drop(locale: Locale) -> &'static str {
    match locale {
        Locale::Fr => "POSER",
        Locale::En => "DROP",
    }
}

pub fn lines_level(locale: Locale, lines: u32, level: u32) -> String {
    match locale {
        Locale::Fr => format!("lignes {lines} / niv {level}"),
        Locale::En => format!("lines {lines} / lv {level}"),
    }
}

pub fn spawn_event(locale: Locale, bell: &str, piece: &str) -> String {
    match locale {
        Locale::Fr => format!("[{bell}] piece {piece}"),
        Locale::En => format!("[{bell}] piece {piece}"),
    }
}

pub fn observe_event(locale: Locale, bits: &str, fx_label: &str) -> String {
    match locale {
        Locale::Fr => format!("observation [{bits}] {fx_label}"),
        Locale::En => format!("observed [{bits}] {fx_label}"),
    }
}

pub fn line_clear_event(locale: Locale, cleared: u32, bits: &str, bonus: u32) -> String {
    match locale {
        Locale::Fr => format!("{cleared} ligne(s) [{bits}] +{bonus}"),
        Locale::En => format!("{cleared} line(s) [{bits}] +{bonus}"),
    }
}

pub fn game_over_event(locale: Locale, score: u32) -> String {
    match locale {
        Locale::Fr => format!("Partie terminee - score {score}"),
        Locale::En => format!("Game over - score {score}"),
    }
}

pub fn retry_hint(locale: Locale) -> &'static str {
    match locale {
        Locale::Fr => "Espace - rejouer",
        Locale::En => "Space - retry",
    }
}

pub fn observe_fx_label(locale: Locale, bits: &str) -> &'static str {
    match (locale, bits) {
        (Locale::Fr, "00") => "effondrement doux",
        (Locale::Fr, "01") => "decalage de phase",
        (Locale::Fr, "10") => "echo de ligne",
        (Locale::Fr, "11") => "mesure forte",
        (Locale::En, "00") => "soft collapse",
        (Locale::En, "01") => "phase shift",
        (Locale::En, "10") => "row echo",
        (Locale::En, "11") => "strong measure",
        (_, _) => "measure",
    }
}

pub fn circuit_heading(locale: Locale) -> &'static str {
    match locale {
        Locale::Fr => "Circuit",
        Locale::En => "Circuit",
    }
}
